File: java/medium/973_k_closest_points_to_origin.java
// https://leetcode.com/problems/kth-largest-element-in-an-array

package java.medium;

import java.util.PriorityQueue;

class KClosestPointsToOrigin {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(
                (a, b) -> Integer.compare(
                        (a[0] * a[0] + a[1] * a[1]),
                        (b[0] * b[0] + b[1] * b[1])));

        for (int[] point : points) {
            minHeap.add(point);
        }

        int[][] result = new int[k][2];
        for (int i = 0; i < k; i++) {
            int[] curr = minHeap.poll();
            result[i][0] = curr[0];
            result[i][1] = curr[1];
        }

        return result;
    }
}

File: go/medium/973_k_closest_points_to_origin.go
// https://leetcode.com/problems/k-closest-points-to-origin/
package medium

import (
	"github.com/emirpasic/gods/queues/priorityqueue"
)

type PointCloset struct {
	idx  int
	dist int
}

func KClosest(points [][]int, k int) [][]int {
	minHeap := priorityqueue.NewWith(func(a, b interface{}) int {
		return b.(PointCloset).dist - a.(PointCloset).dist
	})

	for i, point := range points {
		dist := point[0]*point[0] + point[1]*point[1]
		p := PointCloset{
			idx:  i,
			dist: dist,
		}

		if minHeap.Size() > k {
			minHeap.Dequeue()
		}
		minHeap.Enqueue(p)
	}

	result := make([][]int, 0)
	for i := 0; i < k; i++ {
		p, _ := minHeap.Dequeue()
		idx := p.(PointCloset).idx
		result = append(result, points[idx])
	}

	return result
}
